File: src/commands/cloud_agent/remote_threads.py
"""VM-side conversation metadata without transcript bodies or agent launches.

Claude's versioned SDK owns transcript parsing; Grok's summary.json is its index.
The SDK is installed lazily into an isolated cache only when Claude history exists.
"""

import datetime
import importlib
import json
import os
from pathlib import Path
import shutil
import subprocess
import sys
import tempfile
import sqlite3
import urllib.parse
import urllib.request
import base64

SDK_VERSION = "0.2.152"
RESULT_PREFIX = "RAILWAY-THREADS:"


def read_json(path):
    try:
        with path.open("rb") as file:
            data = file.read(1024 * 1024 + 1)
        if len(data) <= 1024 * 1024:
            return json.loads(data)
    except (OSError, ValueError):
        pass
    return None


def text(value):
    if not isinstance(value, str):
        return ""
    return " ".join("".join(c if c.isprintable() else " " for c in value).split())[:512]


def timestamp(milliseconds):
    if not isinstance(milliseconds, (int, float)):
        return None
    try:
        return datetime.datetime.fromtimestamp(
            milliseconds / 1000, datetime.timezone.utc
        ).isoformat()
    except (ValueError, OverflowError, OSError):
        return None


def iso_timestamp(value):
    if not isinstance(value, str):
        return None
    try:
        date = datetime.datetime.fromisoformat(value.replace("Z", "+00:00"))
        if date.tzinfo is not None:
            return date.astimezone(datetime.timezone.utc).isoformat()
    except ValueError:
        pass
    return None


def process_environment(pid):
    try:
        if not isinstance(pid, int) or pid <= 0:
            return {}
        raw = Path(f"/proc/{pid}/environ").read_bytes()
        return dict(
            entry.decode(errors="replace").split("=", 1)
            for entry in raw.split(b"\0") if b"=" in entry
        )
    except (OSError, ValueError):
        return {}


def alive(pid):
    if not isinstance(pid, int) or pid <= 0:
        return False
    try:
        os.kill(pid, 0)
        return True
    except PermissionError:
        return True
    except OSError:
        return False


def live_identity(pid):
    if not alive(pid):
        return {}
    env = process_environment(pid)
    return {
        "active": True,
        "pane_id": env.get("RAILWAY_THREAD_PANE_ID"),
        "console_name": env.get("RAILWAY_DURABLE_SESSION_NAME"),
    }


def config_roots():
    roots = {
        "claude": {os.environ.get("CLAUDE_CONFIG_DIR", str(Path.home() / ".claude"))},
        "grok": {os.environ.get("GROK_HOME", str(Path.home() / ".grok"))},
    }
    # Sessions launched with per-process config overrides still belong to this VM.
    for process in Path("/proc").glob("[0-9]*"):
        try:
            if process.stat().st_uid != os.getuid():
                continue
            names = [Path(p.decode(errors="replace")).name
                     for p in (process / "cmdline").read_bytes().split(b"\0")[:2]]
            env = process_environment(int(process.name))
            for harness, key in (("claude", "CLAUDE_CONFIG_DIR"), ("grok", "GROK_HOME")):
                if harness in names and env.get(key):
                    roots[harness].add(env[key])
        except OSError:
            continue
    return {h: sorted(Path(p).expanduser() for p in paths) for h, paths in roots.items()}


def claude_sdk():
    cache = Path.home() / ".cache" / "railway" / f"claude-sessions-{SDK_VERSION}"
    if not (cache / "claude_agent_sdk").is_dir():
        import fcntl
        cache.parent.mkdir(parents=True, exist_ok=True)
        with (cache.parent / "claude-sessions.lock").open("w") as lock:
            fcntl.flock(lock, fcntl.LOCK_EX)
            if not (cache / "claude_agent_sdk").is_dir():
                staging = Path(tempfile.mkdtemp(prefix="claude-sessions-", dir=cache.parent))
                try:
                    if shutil.which("uv"):
                        command = ["uv", "pip", "install", "--python", sys.executable]
                    else:
                        command = [sys.executable, "-m", "pip", "install", "--disable-pip-version-check"]
                    subprocess.run(
                        command + ["--target", str(staging), f"claude-agent-sdk=={SDK_VERSION}"],
                        stdin=subprocess.DEVNULL, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                        timeout=90, check=True,
                    )
                    staging.rename(cache)
                finally:
                    if staging.exists():
                        shutil.rmtree(staging)
    sys.path.insert(0, str(cache))
    return importlib.import_module("claude_agent_sdk")


def claude_threads(root, sdk):
    previous = os.environ.get("CLAUDE_CONFIG_DIR")
    os.environ["CLAUDE_CONFIG_DIR"] = str(root)
    try:
        sessions = sdk.list_sessions()
        live = {}
        if shutil.which("claude"):
            try:
                result = subprocess.run(
                    ["claude", "agents", "--json", "--all"], stdin=subprocess.DEVNULL,
                    stdout=subprocess.PIPE, stderr=subprocess.PIPE, timeout=5, check=True,
                )
                for row in json.loads(result.stdout):
                    if isinstance(row, dict) and row.get("sessionId"):
                        old = live.get(row["sessionId"], {})
                        if not old or alive(row.get("pid")):
                            live[row["sessionId"]] = row
            except (OSError, ValueError, subprocess.SubprocessError):
                pass  # Older Claude versions have history but no agents --json.
        rows = []
        for session in sessions:
            if not session.cwd:
                continue  # Resume must run in the saved directory, never an arbitrary /app.
            current = live.get(session.session_id, {})
            state = {"busy": "working", "waiting": "waiting", "idle": "idle"}.get(
                current.get("status"), {"blocked": "waiting", "working": "working",
                                        "failed": "failed", "done": "done"}.get(current.get("state"), "idle")
            )
            rows.append({
                "harness": "claude", "config_dir": str(root),
                "thread": {"id": session.session_id,
                           "title": text(session.custom_title or session.summary or session.first_prompt) or "New Thread",
                           "directory": session.cwd, "created_at": timestamp(session.created_at),
                           "updated_at": timestamp(session.last_modified) or "", "state": state},
                "background_id": current.get("id") if current.get("kind") == "background" else None,
                **live_identity(current.get("pid")),
            })
        return rows
    finally:
        if previous is None:
            os.environ.pop("CLAUDE_CONFIG_DIR", None)
        else:
            os.environ["CLAUDE_CONFIG_DIR"] = previous


def grok_threads(root):
    active = {}
    registry = read_json(root / "active_sessions.json")
    for row in registry if isinstance(registry, list) else []:
        if isinstance(row, dict) and row.get("session_id") and alive(row.get("pid")):
            active[row["session_id"]] = live_identity(row["pid"])
    rows = []
    for path in (root / "sessions").glob("*/*/summary.json"):
        row = read_json(path)
        if not isinstance(row, dict):
            continue
        info = row.get("info") or {}
        if not isinstance(info, dict) or not isinstance(info.get("id"), str) or not isinstance(info.get("cwd"), str):
            continue
        if not info["id"] or not info["cwd"].startswith("/"):
            continue
        kind = text(row.get("session_kind"))
        hidden = row.get("hidden")
        if hidden is True or (hidden is None and kind.startswith("subagent")):
            continue
        title = text(row.get("generated_title")) or text(row.get("session_summary"))
        if not title and not row.get("num_messages") and not (
            kind == "fork" or row.get("parent_session_id") or row.get("forked_at")
        ):
            continue
        rows.append({
            "harness": "grok", "config_dir": str(root),
            "thread": {"id": info["id"], "directory": info["cwd"],
                       "title": title or "New Thread", "created_at": iso_timestamp(row.get("created_at")),
                       "updated_at": iso_timestamp(row.get("last_active_at")) or iso_timestamp(row.get("updated_at")) or "",
                       "state": "idle"},
            **active.get(info["id"], {}),
        })
    return rows


def codex_threads():
    root = Path(os.environ.get("CODEX_HOME", str(Path.home() / ".codex")))
    paths = sorted((p for p in root.glob("state_*.sqlite") if p.stem[6:].isdigit()),
                   key=lambda p: int(p.stem[6:]), reverse=True)
    if not paths:
        return []
    # Codex maintains a metadata index independently of App Server's lifetime.
    with sqlite3.connect(paths[0].absolute().as_uri() + "?mode=ro", uri=True, timeout=2) as db:
        db.row_factory = sqlite3.Row
        columns = {row[1] for row in db.execute("PRAGMA table_info(threads)")}
        required = {"id", "cwd", "title", "created_at", "updated_at", "archived", "source"}
        if not required.issubset(columns):
            raise ValueError("Unsupported Codex metadata schema")
        fields = sorted(required | ({"name", "preview"} & columns))
        rows = db.execute("SELECT " + ",".join(fields) + " FROM threads WHERE archived=0 ORDER BY updated_at DESC")
        return [{"harness": "codex", "config_dir": str(root), "thread": {
            "id": row["id"], "directory": row["cwd"],
            "title": text((row["name"] if "name" in columns else None) or row["title"] or
                          (row["preview"] if "preview" in columns else None)) or "New Thread",
            "created_at": timestamp(row["created_at"] * 1000),
            "updated_at": timestamp(row["updated_at"] * 1000) or "", "state": "idle",
        }} for row in rows if row["source"] in ("cli", "vscode", "appServer", "exec")]


def opencode_threads():
    data = Path(os.environ.get("XDG_DATA_HOME", str(Path.home() / ".local" / "share"))) / "opencode"
    override = os.environ.get("OPENCODE_DB")
    paths = [Path(override) if Path(override).is_absolute() else data / override] if override and override != ":memory:" else sorted(data.glob("*.db"))
    rows = []
    for path in paths:
        if not path.is_file():
            continue
        with sqlite3.connect(path.absolute().as_uri() + "?mode=ro", uri=True, timeout=2) as db:
            db.row_factory = sqlite3.Row
            for table, harness in (("session", "opencode"), ("session_v2", "opencode2")):
                columns = {row[1] for row in db.execute(f"PRAGMA table_info({table})")}
                if not columns:
                    continue
                if not {"id", "title", "directory", "parent_id", "time_created", "time_updated", "time_archived"}.issubset(columns):
                    raise ValueError("Unsupported OpenCode metadata schema")
                for row in db.execute(f"SELECT id,title,directory,time_created,time_updated FROM {table} WHERE parent_id IS NULL AND time_archived IS NULL ORDER BY time_updated DESC"):
                    title = text(row["title"])
                    rows.append({"harness": harness, "config_dir": str(Path.home() / ".config" / "opencode"),
                        "database": str(path), "thread": {
                            "id": row["id"], "directory": row["directory"],
                            "title": "New Thread" if not title or title.startswith("New session - ") else title,
                            "created_at": timestamp(row["time_created"]), "updated_at": timestamp(row["time_updated"]) or "", "state": "idle",
                        }})
    return rows if paths else opencode_server_threads()


def opencode_server_threads():
    root = Path.home() / ".railway" / "desktop" / "opencode"
    state = read_json(root / "server.json")
    if not isinstance(state, dict):
        return []
    harness = state.get("harness", "opencode")
    if harness not in ("opencode", "opencode2"):
        return []
    beta = harness == "opencode2"
    credentials = base64.b64encode((state["username"] + ":" + state["password"]).encode()).decode()
    rows, cursor, seen = [], None, set()
    while True:
        query = {"limit": 100, "order": "desc"} if beta else {"directory": state["directory"]}
        if cursor:
            query["cursor"] = cursor
        request = urllib.request.Request("http://127.0.0.1:8080/" + ("api/" if beta else "") +
                                         "session?" + urllib.parse.urlencode(query),
                                         headers={"Authorization": "Basic " + credentials})
        with urllib.request.build_opener(urllib.request.ProxyHandler({})).open(request, timeout=5) as response:
            page = json.load(response)
        for row in page["data"] if beta else page:
            if row.get("parentID") or (row.get("time") or {}).get("archived"):
                continue
            title = text(row.get("title"))
            rows.append({"harness": harness, "config_dir": str(root), "thread": {
                "id": row["id"], "title": "New Thread" if not title or title.startswith("New session - ") else title,
                "directory": (row.get("location") or {}).get("directory") or row.get("directory") or state["directory"],
                "created_at": timestamp(row["time"]["created"]),
                "updated_at": timestamp(row["time"]["updated"]) or "", "state": "idle",
            }})
        cursor = (page.get("cursor") or {}).get("next") if beta else None
        if not cursor:
            return rows
        if cursor in seen:
            raise ValueError("Repeated conversation cursor")
        seen.add(cursor)


def discover():
    roots = config_roots()
    rows, warnings, failed = [], [], []
    for harness in ("grok", "claude"):
        try:
            for root in roots[harness]:
                if harness == "grok":
                    rows.extend(grok_threads(root))
                elif (root / "projects").is_dir():
                    rows.extend(claude_threads(root, claude_sdk()))
        except Exception as error:
            failed.append(harness)
            warnings.append(f"{harness} history unavailable ({type(error).__name__})")
    for harnesses, read in ((["codex"], codex_threads), (["opencode", "opencode2"], opencode_threads)):
        try:
            rows.extend(read())
        except Exception as error:
            failed.extend(harnesses)
            warnings.append(f"{'/'.join(harnesses)} history unavailable ({type(error).__name__})")
    # Relocations and restored histories can leave duplicate IDs in the tree.
    newest = {}
    for row in sorted(rows, key=lambda r: r["thread"]["updated_at"], reverse=True):
        newest.setdefault((row["harness"], row["thread"]["id"]), row)
    return {"threads": list(newest.values()), "warnings": warnings, "failed": failed}


if __name__ == "__main__":
    print(RESULT_PREFIX + json.dumps(discover(), ensure_ascii=True))

File: src/commands/cloud_agent/remote_threads.rs
//! Persisted harness conversations discovered on the VM over one bounded SSH call.
use std::{process::Stdio, time::Duration};

use anyhow::{Context, Result, bail};
use tokio::io::AsyncWriteExt;

use super::client_sessions::{self, Thread};
use crate::commands::{code, ssh::native};

const HELPER: &str = include_str!("remote_threads.py");
const PREFIX: &str = "RAILWAY-THREADS:";

#[derive(Clone, Debug, serde::Deserialize)]
pub(crate) struct RemoteThread {
    pub harness: String,
    pub thread: Thread,
    pub config_dir: String,
    #[serde(default)]
    pub active: bool,
    pub pane_id: Option<String>,
    pub console_name: Option<String>,
    pub background_id: Option<String>,
    pub database: Option<String>,
}

impl RemoteThread {
    pub fn name(&self, agent_id: &str) -> String {
        client_sessions::name(&self.harness, agent_id, Some(&self.thread.id))
    }

    pub fn resume_command(&self, pane_id: &str) -> Result<String> {
        client_sessions::validate_id(&self.thread.id)?;
        let variable = match self.harness.as_str() {
            "claude" => Some("CLAUDE_CONFIG_DIR"),
            "grok" => Some("GROK_HOME"),
            "codex" => Some("CODEX_HOME"),
            "opencode" | "opencode2" | "railway" => None,
            _ => bail!("Unsupported remote conversation harness"),
        };
        if !self.thread.directory.starts_with('/') || !self.config_dir.starts_with('/') {
            bail!("Conversation has no absolute working/config directory");
        }
        let invocation = match self
            .background_id
            .as_deref()
            .filter(|_| self.harness == "claude")
        {
            Some(id) => vec!["claude".into(), "attach".into(), id.into()],
            None if self.harness == "codex" => vec![
                "codex".into(),
                "--ask-for-approval".into(),
                "never".into(),
                "--sandbox".into(),
                "danger-full-access".into(),
                "resume".into(),
                self.thread.id.clone(),
            ],
            None if self.harness.starts_with("opencode") => {
                let mut args = vec![self.harness.clone()];
                if self.harness == "opencode2" {
                    args.extend(["--standalone".into(), "--auto".into()]);
                }
                args.extend(["--session".into(), self.thread.id.clone()]);
                args
            }
            None if self.harness == "railway" => vec![
                "railway-agent-tui".into(),
                "--session".into(),
                self.thread.id.clone(),
            ],
            None => vec![
                self.harness.clone(),
                "--resume".into(),
                self.thread.id.clone(),
            ],
        };
        let config = format!(
            "{}{}",
            variable
                .map(|variable| format!("export {variable}={}; ", quote(&self.config_dir)))
                .unwrap_or_default(),
            self.database
                .as_ref()
                .filter(|_| self.harness == "opencode2")
                .map(|path| format!("export OPENCODE_DB={}; ", quote(path)))
                .unwrap_or_default()
        );
        Ok(format!(
            "{}export RAILWAY_CODE_AUTOSTARTED=1; export RAILWAY_THREAD_PANE_ID={}; {config}cd -- {} && exec {}",
            code::harness_env_prefix(),
            quote(pane_id),
            quote(&self.thread.directory),
            crate::util::shell::shell_join(&invocation),
        ))
    }
}

fn quote(value: &str) -> String {
    crate::util::shell::shell_join(&[value.into()])
}

#[derive(Default, serde::Deserialize)]
pub(crate) struct Discovery {
    pub threads: Vec<RemoteThread>,
    pub warnings: Vec<String>,
    pub failed: Vec<String>,
}

pub(crate) async fn discover(info: &code::ConnectInfo) -> Result<Discovery> {
    let output = run_helper(info, "").await?;
    let mut result: Discovery =
        serde_json::from_str(&output).context("Invalid conversation inventory")?;
    result.threads.retain(|row| {
        matches!(
            row.harness.as_str(),
            "claude" | "grok" | "codex" | "opencode" | "opencode2" | "railway"
        ) && client_sessions::validate_id(&row.thread.id).is_ok()
    });
    Ok(result)
}

pub(crate) async fn delete(
    info: &code::ConnectInfo,
    harness: &str,
    id: &str,
    consoles: &[String],
) -> Result<()> {
    client_sessions::validate_id(id)?;
    let request =
        serde_json::json!({"harness": harness, "id": id, "consoles": consoles}).to_string();
    let result: serde_json::Value =
        serde_json::from_str(&run_helper(info, &quote(&request)).await?)?;
    anyhow::ensure!(
        result["deleted"] == id,
        "The harness did not confirm conversation deletion"
    );
    Ok(())
}

async fn run_helper(info: &code::ConnectInfo, argument: &str) -> Result<String> {
    let mut command = tokio::process::Command::new("ssh");
    command
        .args(native::relay_port_args())
        .args(&info.relay_opts);
    if let Some(identity) = &info.identity {
        command.arg("-i").arg(identity);
    }
    // Read-only metadata runs never request a durable console or a PTY.
    let remote = format!(
        "{}export RAILWAY_CODE_AUTOSTARTED=1 RAILWAY_THREAD_DISCOVERY=1; python3 - {argument}",
        code::harness_env_prefix()
    );
    let mut child = command
        .args(["-T", "-o", "BatchMode=yes", "-o", "ConnectTimeout=10", "--"])
        .arg(native::relay_destination(&info.ssh_target))
        .arg(remote)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .kill_on_drop(true)
        .spawn()
        .context("Starting conversation command over SSH")?;
    let mut stdin = child.stdin.take().context("SSH stdin unavailable")?;
    stdin.write_all(HELPER.as_bytes()).await?;
    drop(stdin);
    let output = tokio::time::timeout(Duration::from_secs(120), child.wait_with_output())
        .await
        .context("Conversation command timed out")??;
    if !output.status.success() {
        bail!(
            "Conversation command failed: {}",
            String::from_utf8_lossy(&output.stderr).trim()
        );
    }
    let stdout = String::from_utf8_lossy(&output.stdout);
    let json = stdout
        .lines()
        .find_map(|line| line.strip_prefix(PREFIX))
        .context("SSH returned no conversation inventory")?;
    Ok(json.to_owned())
}

#[cfg(test)]
pub(crate) mod tests {
    use super::*;

    pub(crate) fn thread(harness: &str, id: &str) -> RemoteThread {
        RemoteThread {
            harness: harness.into(),
            thread: Thread {
                id: id.into(),
                title: "Saved conversation".into(),
                directory: "/app/it's a project".into(),
                created_at: None,
                updated_at: "2026-09-10T10:00:00Z".into(),
                state: "idle".into(),
            },
            config_dir: "/home/user/custom config".into(),
            active: false,
            pane_id: None,
            console_name: None,
            background_id: None,
            database: None,
        }
    }

    #[test]
    fn resumes_exact_thread_in_original_directory_and_config() {
        for harness in ["claude", "grok"] {
            let thread = thread(harness, "saved-id");
            let command = thread.resume_command("pane-id").unwrap();
            let words = shlex::split(&command).unwrap();
            let cd = words.iter().position(|w| w == "cd").unwrap();
            assert_eq!(
                &words[cd..],
                &[
                    "cd",
                    "--",
                    "/app/it's a project",
                    "&&",
                    "exec",
                    harness,
                    "--resume",
                    "saved-id"
                ]
            );
            assert!(
                words
                    .iter()
                    .any(|w| w.ends_with("=/home/user/custom config;"))
            );
        }
        let mut thread = thread("claude", "saved-id");
        thread.background_id = Some("job-id".into());
        assert!(
            thread
                .resume_command("pane-id")
                .unwrap()
                .ends_with("exec claude attach job-id")
        );
        thread.thread.directory.clear();
        assert!(thread.resume_command("pane-id").is_err());
    }

    #[test]
    fn vm_metadata_discovery_fixtures() {
        let output = std::process::Command::new("python3")
            .arg("src/commands/cloud_agent/remote_threads_test.py")
            .env("PYTHONDONTWRITEBYTECODE", "1")
            .output()
            .expect("python3 is required for VM metadata integration tests");
        assert!(
            output.status.success(),
            "{}",
            String::from_utf8_lossy(&output.stderr)
        );
    }
}
